### PR TITLE
In Inductor, be willing to generate deferred runtime asserts when unbacked

### PR DESCRIPTION
**Pull request by laithsakka**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #138804

Title + we avoid calling defer_assert when we statically know the guard results. 
timing for pnasnet5large

```
TIMING: code_gen:21.79672 inductor_compile:39.57726 backend_compile:65.30649 entire_frame_compile:95.22052 total_wall_time:95.22052
```
matches with out the diff
```
TIMING: code_gen:21.89314 inductor_compile:39.72298 backend_compile:65.38539 entire_frame_compile:95.0854 total_wall_time:95.0854
```

cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov @rec

### CODE REVIEW

File: torch/_inductor/sizevars.py [hunk 1]
```diff
@@ -890,9 +895,9 @@ class SimplifyIndexing(V.WrapperHandler):  # type: ignore[name-defined]
     def __init__(self, inner, var_ranges: VarRanges) -> None:
         super().__init__(inner)
         self.name = "SimplifyIndexing"
-        self._simplify: Callable[
-            [Expr], Expr
-        ] = lambda index: V.graph.sizevars.simplify_with_ranges(index, var_ranges)
+        self._simplify: Callable[[Expr], Expr] = (
+            lambda index: V.graph.sizevars.simplify_with_ranges(index, var_ranges)
```
Comment (ezyang) [inline torch/_inductor/sizevars.py:899]: this is... no op?

**Review by ezyang — APPROVED** [review]

File: torch/_inductor/sizevars.py (hunk 1 above)
Comment (laithsakka) [reply torch/_inductor/sizevars.py:899]: my vs code does some annoying auto format

**Review by ezyang — APPROVED** [review]

**Review by ezyang — APPROVED** [review]

**Comment by laithsakka** [comment]
@pytrochbot merge

**Comment by laithsakka** [comment]
add the unit test

**Comment by huydhn** [comment]
@pytorchbot revert -m 'Sorry for reverting your change, but it seems to fail pr_time_benchmarks job in trunk' -c nosignal

[GH job link](https://github.com/pytorch/pytorch/actions/runs/11532781856/job/32105419448) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/f1a677cba5ef7514f2cf303753d3117528867a33)

**Comment by laithsakka** [comment]
ahh probably land time conflict since i enabled regression detection on update_hint_benchmark over the weekend will rebase and fix

**Pull request closed without merging**